File: backend/tools/check_comments.py
"""Prüft Kommentarregeln in ``app`` und ``tools``."""

import hashlib
import io
import json
import re
import sys
import tokenize
from dataclasses import dataclass
from pathlib import Path
from typing import Final, cast

ROOT: Final = Path(__file__).resolve().parent.parent
ALLOW_PATH: Final = Path(__file__).resolve().parent / "lint-allow.json"

BLOCK_MAX_LINES: Final = 2
DOC_MAX_LINES: Final = 3

YEAR: Final = re.compile(r"20\d{2}(?:-\d{2}-\d{2})?")
PR_NUMBER: Final = re.compile(r"#\d+|\bPR\s*\d+\b", re.IGNORECASE)
FORBIDDEN_WORDS: Final = re.compile(
    r"(?<![A-Za-zÄÖÜäöüß])(bis|frueher|früher|jetzt|seit)(?![A-Za-zÄÖÜäöüß])",
    re.IGNORECASE,
)


@dataclass(frozen=True, slots=True)
class RawComment:
    """Ein Kommentar oder Docstring mit Zeilenspanne."""

    kind: str
    start: int
    end: int
    text: str
    standalone: bool


@dataclass(frozen=True, slots=True)
class Violation:
    """Ein gemeldeter Verstoß mit Ort, Regel und Grund."""

    path: str
    line: int
    rule: str
    text: str
    reason: str

    @property
    def key(self) -> str:
        """Schlüssel für die Ausnahmeliste: Pfad und Text, nicht die Zeile."""
        stamp = re.sub(r"\s+", " ", f"{self.rule} {self.text}").strip().encode()
        return f"{self.path}#{hashlib.sha256(stamp).hexdigest()[:8]}"


def forbidden_reason(text: str) -> str | None:
    """Findet den Grund, warum ein Kommentartext verboten ist."""
    if YEAR.search(text):
        return "Jahreszahl oder Datum im Kommentar"
    if PR_NUMBER.search(text):
        return "PR-Nummer im Kommentar"
    word = FORBIDDEN_WORDS.search(text)
    if word:
        return f"verbotenes Wort „{word.group(1)}“"
    return None


def scan_source(text: str) -> list[RawComment]:
    """Zerlegt Quelltext in Kommentare und Docstrings via ``tokenize``."""
    found: list[RawComment] = []
    for tok in tokenize.generate_tokens(io.StringIO(text).readline):
        if tok.type == tokenize.COMMENT:
            before = tok.line[: tok.start[1]]
            found.append(
                RawComment(
                    kind="line",
                    start=tok.start[0],
                    end=tok.start[0],
                    text=tok.string.removeprefix("#"),
                    standalone=before.strip() == "",
                ),
            )
        elif tok.type == tokenize.STRING and (
            tok.string.startswith('"""') or tok.string.startswith("'''")
        ):
            found.append(
                RawComment(
                    kind="doc",
                    start=tok.start[0],
                    end=tok.end[0],
                    text=tok.string[3:-3],
                    standalone=True,
                ),
            )
    return found


def standalone_runs(comments: list[RawComment]) -> list[list[RawComment]]:
    """Gruppiert zusammenhängende, alleinstehende Zeilenkommentare."""
    runs: list[list[RawComment]] = []
    run: list[RawComment] = []
    for comment in comments:
        if comment.kind == "line" and comment.standalone:
            if run and comment.start == run[-1].start + 1:
                run.append(comment)
            else:
                if run:
                    runs.append(run)
                run = [comment]
        else:
            if run:
                runs.append(run)
            run = []
    if run:
        runs.append(run)
    return runs


def run_violations(path: str, runs: list[list[RawComment]]) -> list[Violation]:
    """Meldet Kommentarläufe über der Zeilengrenze."""
    return [
        Violation(
            path,
            r[0].start,
            "block",
            " ".join(c.text for c in r),
            f"{len(r)} zusammenhängende Zeilen",
        )
        for r in runs
        if len(r) > BLOCK_MAX_LINES
    ]


def content_violation(path: str, comment: RawComment) -> Violation | None:
    """Prüft einen Kommentar auf Länge und verbotenen Inhalt."""
    reason = forbidden_reason(comment.text)
    if reason:
        return Violation(path, comment.start, "inhalt", comment.text, reason)
    span = comment.end - comment.start + 1
    if comment.kind == "doc" and span > DOC_MAX_LINES:
        return Violation(
            path,
            comment.start,
            "docstring",
            comment.text,
            f"{span} Zeilen, Grenze {DOC_MAX_LINES}",
        )
    return None


def group_violations(path: str, comments: list[RawComment]) -> list[Violation]:
    """Bildet aus den Rohfunden die Verstöße einer Datei."""
    violations = run_violations(path, standalone_runs(comments))
    violations += [v for c in comments if (v := content_violation(path, c)) is not None]
    return violations


def collect_files() -> list[Path]:
    """Listet die Python-Dateien unter ``app`` und ``tools``."""
    files = {*(ROOT / "app").rglob("*.py"), *(ROOT / "tools").rglob("*.py")}
    return sorted(files)


def all_violations() -> list[Violation]:
    """Sammelt die Verstöße über alle geprüften Dateien."""
    found: list[Violation] = []
    for file in collect_files():
        text = file.read_text(encoding="utf-8")
        rel = file.relative_to(ROOT).as_posix()
        found.extend(group_violations(rel, scan_source(text)))
    return found


def read_allow() -> dict[str, list[str]]:
    """Liest die Ausnahmeliste, leer wenn die Datei fehlt."""
    try:
        raw = ALLOW_PATH.read_text(encoding="utf-8")
    except FileNotFoundError:
        return {}
    return cast("dict[str, list[str]]", json.loads(raw))


def write_allow(key: str, keys: list[str]) -> None:
    """Schreibt die Ausnahmeliste neu, andere Schlüssel bleiben stehen."""
    allow = read_allow()
    allow[key] = keys
    ALLOW_PATH.write_text(json.dumps(allow, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")


def main() -> None:
    """Prüft, meldet neue Verstöße oder schreibt die Ausnahmeliste neu."""
    violations = all_violations()
    keys = sorted({v.key for v in violations})
    if "--write-allow" in sys.argv:
        write_allow("comments", keys)
        print(f"{len(keys)} Ausnahmen für comments geschrieben.")
        return
    allowed = set(read_allow().get("comments", []))
    reported = [v for v in violations if v.key not in allowed]
    for v in reported:
        print(f"{v.path}:{v.line}  {v.rule}  {v.reason}")
    if reported:
        print(f"Neue Kommentarverstöße: {len(reported)}.")
        raise SystemExit(1)
    print("Keine neuen Kommentarverstöße.")


if __name__ == "__main__":
    main()

File: frontend/tools/check-comments.mjs
#!/usr/bin/env node
/**
 * Prüft Kommentarregeln in `src` und `tools`.
 * Kommentare sagen Warum, nie Was.
 */
import { readFileSync, readdirSync, statSync, writeFileSync } from 'node:fs';
import { createHash } from 'node:crypto';
import { join, relative } from 'node:path';
import { fileURLToPath } from 'node:url';

const YEAR = /20\d{2}(?:-\d{2}-\d{2})?/;
const PR_NUMBER = /#\d+|\bPR\s*\d+\b/i;
const WORD_BOUNDARY = '(?<![A-Za-zÄÖÜäöüß])';
const WORD_BOUNDARY_END = '(?![A-Za-zÄÖÜäöüß])';
const FORBIDDEN_WORDS = new RegExp(
  `${WORD_BOUNDARY}(bis|frueher|früher|jetzt|seit)${WORD_BOUNDARY_END}`,
  'i',
);

/** Findet den Grund, warum ein Kommentartext verboten ist. */
function forbiddenReason(text) {
  if (YEAR.test(text)) return 'Jahreszahl oder Datum im Kommentar';
  if (PR_NUMBER.test(text)) return 'PR-Nummer im Kommentar';
  const word = FORBIDDEN_WORDS.exec(text);
  if (word) return `verbotenes Wort „${word[1]}“`;
  return null;
}

function lineAt(source, index) {
  let line = 1;
  for (let i = 0; i < index; i += 1) if (source[i] === '\n') line += 1;
  return line;
}

/** Zerlegt Quelltext in Zeilen- und Blockkommentare. Strings bleiben aussen vor. */
function scanCode(source) {
  const found = [];
  let i = 0;
  while (i < source.length) {
    const ch = source[i];
    if (ch === '"' || ch === "'" || ch === '`') {
      let j = i + 1;
      while (j < source.length && source[j] !== ch) j += source[j] === '\\' ? 2 : 1;
      i = j + 1;
      continue;
    }
    if (ch === '/' && source[i + 1] === '/') {
      const end = source.indexOf('\n', i);
      const stop = end === -1 ? source.length : end;
      const before = source.slice(source.lastIndexOf('\n', i) + 1, i);
      found.push({
        type: 'line',
        standalone: before.trim() === '',
        line: lineAt(source, i),
        text: source.slice(i + 2, stop),
      });
      i = stop;
      continue;
    }
    if (ch === '/' && source[i + 1] === '*') {
      const end = source.indexOf('*/', i + 2);
      const stop = end === -1 ? source.length : end + 2;
      const isDoc = source[i + 2] === '*' && source[i + 3] !== '/';
      found.push({
        type: isDoc ? 'doc' : 'block',
        line: lineAt(source, i),
        endLine: lineAt(source, stop),
        text: source.slice(i + (isDoc ? 3 : 2), stop - 2),
      });
      i = stop;
      continue;
    }
    i += 1;
  }
  return found;
}

/** Zerlegt eine Vorlage in `<!-- -->`-Kommentare. */
function scanHtml(source) {
  const found = [];
  const pattern = /<!--([\s\S]*?)-->/g;
  for (const match of source.matchAll(pattern)) {
    const start = match.index;
    const stop = start + match[0].length;
    found.push({ type: 'html', line: lineAt(source, start), endLine: lineAt(source, stop), text: match[1] });
  }
  return found;
}

/** Baut aus den Rohtreffern die Verstöße einer Datei. */
function violationsIn(path, comments) {
  const found = [];
  const lineRuns = [];
  let run = [];
  for (const comment of comments) {
    if (comment.type === 'line' && comment.standalone) {
      if (run.length > 0 && comment.line === run[run.length - 1] + 1) run.push(comment.line);
      else {
        if (run.length > 0) lineRuns.push(run);
        run = [comment.line];
      }
    } else if (run.length > 0) {
      lineRuns.push(run);
      run = [];
    }
  }
  if (run.length > 0) lineRuns.push(run);
  for (const r of lineRuns) {
    if (r.length > 2)
      found.push({
        line: r[0],
        rule: 'block',
        text: comments
          .filter((c) => r.includes(c.line))
          .map((c) => c.text)
          .join(' '),
        reason: `${r.length} zusammenhängende Zeilen`,
      });
  }
  for (const comment of comments) {
    const reason = forbiddenReason(comment.text);
    if (reason) found.push({ line: comment.line, rule: 'inhalt', text: comment.text, reason });
    if (comment.type === 'doc') {
      const span = comment.endLine - comment.line + 1;
      if (span > 3) {
        found.push({
          line: comment.line,
          rule: 'docstring',
          text: comment.text,
          reason: `${span} Zeilen, Grenze 3`,
        });
      }
    }
    if (comment.type === 'block' || comment.type === 'html') {
      const span = comment.endLine - comment.line + 1;
      if (span > 2) {
        found.push({
          line: comment.line,
          rule: 'block',
          text: comment.text,
          reason: `${span} Zeilen, Grenze 2`,
        });
      }
    }
  }
  return found.map((v) => ({ path, ...v }));
}

function collectFiles(folder, extensions, found) {
  for (const name of readdirSync(folder)) {
    const path = join(folder, name);
    if (statSync(path).isDirectory()) {
      collectFiles(path, extensions, found);
      continue;
    }
    if (extensions.some((ext) => name.endsWith(ext))) found.push(path);
  }
  return found;
}

/** Sucht Kommentarverstöße unter `root/src/app` und `root/tools`. */
export function findViolations(root) {
  const files = [
    ...collectFiles(join(root, 'src', 'app'), ['.ts', '.html', '.scss'], []),
    ...collectFiles(join(root, 'tools'), ['.mjs'], []),
  ];
  const found = [];
  for (const file of files) {
    const source = readFileSync(file, 'utf8');
    const path = relative(root, file);
    const comments = file.endsWith('.html') ? scanHtml(source) : scanCode(source);
    found.push(...violationsIn(path, comments));
  }
  return found;
}

/** Der Schlüssel hängt am Text, nicht an der Zeile. Eine Verschiebung zählt nicht. */
export function allowKey(v) {
  const text = `${v.rule} ${v.text}`.replace(/\s+/g, ' ').trim();
  return `${v.path}#${createHash('sha256').update(text).digest('hex').slice(0, 8)}`;
}

function readAllow(allowPath) {
  try {
    return JSON.parse(readFileSync(allowPath, 'utf8'));
  } catch {
    return {};
  }
}

/** Meldet Verstöße unter `root` ohne die in `allowPath` freigegebenen. */
export function report(root, allowPath) {
  const allowed = new Set(readAllow(allowPath).comments ?? []);
  return findViolations(root).filter((v) => !allowed.has(allowKey(v)));
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const ROOT = fileURLToPath(new URL('..', import.meta.url));
  const ALLOW_PATH = join(ROOT, 'tools', 'lint-allow.json');

  if (process.argv.includes('--write-allow')) {
    const keys = [...new Set(findViolations(ROOT).map(allowKey))].sort();
    const allow = readAllow(ALLOW_PATH);
    allow.comments = keys;
    writeFileSync(ALLOW_PATH, JSON.stringify(allow, null, 2) + '\n');
    console.log(`${keys.length} Ausnahmen für comments geschrieben.`);
    process.exit(0);
  }

  const reported = report(ROOT, ALLOW_PATH);
  for (const v of reported) console.log(`${v.path}:${v.line}  ${v.rule}  ${v.reason}`);
  if (reported.length > 0) {
    console.error(`Neue Kommentarverstöße: ${reported.length}.`);
    process.exit(1);
  }
  console.log('Keine neuen Kommentarverstöße.');
}
